dms_version/models/version.py

```diff
@@ -0,0 +1,185 @@
+        related="storage_id.incremental_versions", readonly=True
+    )
+
+    previous_version_id = fields.Many2one(
+        comodel_name="dms.version", ondelete="set null", string="Previous Version"
+    )
+
+    next_version_id = fields.Many2one(
+        comodel_name="dms.version", ondelete="set null", string="Next Version"
+    )
+
+    is_compress = fields.Boolean(string="Compressed", default=False)
+
+    is_incremental = fields.Boolean(string="Incremental", default=False)
+
+    size = fields.Integer(string="Size", readonly=True)
+
+    content = fields.Binary(
+        compute="_compute_content",
+        inverse="_inverse_content",
+        string="Content",
+        attachment=False,
+        prefetch=False,
+        required=True,
+        store=False,
+    )
+
+    binary = fields.Binary(
+        string="Binary", prefetch=False, attachment=True, invisible=True
+    )
+
+    def get_binary(self, patch=True):
+        self.ensure_one()
+        binary = base64.b64decode(self.binary)
+        if self.is_compress:
+            binary = zlib.decompress(binary)
+        if self.is_incremental and patch:
+            prev_version = self.previous_version_id
+            prev_binary = prev_version.get_binary()
+            binary = bsdiff4.patch(prev_binary, binary)
+        return binary
+
+    @api.model
+    def clean_versions(self, files, max_size=None):
+        for file in files.filtered(lambda rec: rec.count_versions > 0):
+            max_size = max_size or file.storage_id.stored_versions
+            if file.count_versions > max_size:
+                remove_size = file.count_versions - max_size
+                remove_versions = self.env["dms.version"]
+                current = file.version_ids.filtered(
+                    lambda rec: not rec.previous_version_id
+                )
+                while len(remove_versions) < remove_size:
+                    remove_versions |= current
+                    current = current.next_version_id
+                current.write(
+                    {
+                        "content": current.content,
+                        "previous_version_id": None,
+                        "is_incremental": False,
+                    }
+                )
+                remove_versions.unlink()
+
+    @api.model
+    def add_version(self, file, content):
+        domain = [
+            ("file_id", "=", file.id),
+            ("next_version_id", "=", False),
+        ]
+        record = self.search(domain, limit=1)
+        if not record:
+            return self.create(
+                {"name": file.name, "file_id": file.id, "content": content}
+            )
+        new_version = self.create(
+            {
+                "name": file.name,
+                "file_id": file.id,
+                "content": content,
+                "previous_version_id": record.id,
+            }
+        )
+        record.write({"next_version_id": new_version.id})
+        return new_version
```

Comment:
IMO always create two versions and maybe other similar options is something like this

```suggestion
        domain = [
            ("file_id", "=", file.id),
        ]
        record = self.search(domain, order='id desc' limit=1)
        if not record:
            return self.create(
                {"name": file.name, "file_id": file.id, "content": content}
            )
        else:        
            new_version = self.create(
                {
                    "name": file.name,
                    "file_id": file.id,
                    "content": content,
                    "previous_version_id": record.id,
                }
            )
            record.write({"next_version_id": new_version.id})
            return new_version
```